File: jscomp/test/bs_link_list_test.js
'use strict';

var Mt         = require("./mt.js");
var Block      = require("../../lib/js/block.js");
var Bs_List    = require("../../lib/js/bs_List.js");
var Bs_Array   = require("../../lib/js/bs_Array.js");
var Caml_int32 = require("../../lib/js/caml_int32.js");

var suites = [/* [] */0];

var test_id = [0];

function eq(loc, x, y) {
  test_id[0] = test_id[0] + 1 | 0;
  suites[0] = /* :: */[
    /* tuple */[
      loc + (" id " + test_id[0]),
      (function () {
          return /* Eq */Block.__(0, [
                    x,
                    y
                  ]);
        })
    ],
    suites[0]
  ];
  return /* () */0;
}

function b(loc, x) {
  test_id[0] = test_id[0] + 1 | 0;
  suites[0] = /* :: */[
    /* tuple */[
      loc + (" id " + test_id[0]),
      (function () {
          return /* Ok */Block.__(4, [x]);
        })
    ],
    suites[0]
  ];
  return /* () */0;
}

var u = Bs_List.init(5, (function (i) {
        return Caml_int32.imul(i, i);
      }));

function f(i) {
  return eq("File \"bs_link_list_test.ml\", line 23, characters 7-14", Bs_List.nthAssert(u, i), Caml_int32.imul(i, i));
}

for(var i = 0; i <= 4; ++i){
  f(i);
}

eq("File \"bs_link_list_test.ml\", line 27, characters 5-12", Bs_List.map(u, (function (i) {
            return i + 1 | 0;
          })), /* :: */[
      1,
      /* :: */[
        2,
        /* :: */[
          5,
          /* :: */[
            10,
            /* :: */[
              17,
              /* [] */0
            ]
          ]
        ]
      ]
    ]);

eq("File \"bs_link_list_test.ml\", line 30, characters 5-12", Bs_List.flatten(/* :: */[
          /* :: */[
            1,
            /* [] */0
          ],
          /* :: */[
            /* :: */[
              2,
              /* [] */0
            ],
            /* :: */[
              /* :: */[
                3,
                /* [] */0
              ],
              /* :: */[
                /* [] */0,
                /* :: */[
                  Bs_List.init(4, (function (i) {
                          return i;
                        })),
                  /* [] */0
                ]
              ]
            ]
          ]
        ]), /* :: */[
      1,
      /* :: */[
        2,
        /* :: */[
          3,
          /* :: */[
            0,
            /* :: */[
              1,
              /* :: */[
                2,
                /* :: */[
                  3,
                  /* [] */0
                ]
              ]
            ]
          ]
        ]
      ]
    ]);

eq("File \"bs_link_list_test.ml\", line 37, characters 5-12", Bs_List.toArray(Bs_List.append(Bs_List.init(100, (function (i) {
                    return i;
                  })), Bs_List.init(100, (function (i) {
                    return i;
                  })))), Bs_Array.append(Bs_Array.init(100, (function (i) {
                return i;
              })), Bs_Array.init(100, (function (i) {
                return i;
              }))));

eq("ZIP", Bs_List.zip(/* :: */[
          1,
          /* :: */[
            2,
            /* :: */[
              3,
              /* [] */0
            ]
          ]
        ], /* :: */[
          3,
          /* :: */[
            4,
            /* [] */0
          ]
        ]), /* :: */[
      /* tuple */[
        1,
        3
      ],
      /* :: */[
        /* tuple */[
          2,
          4
        ],
        /* [] */0
      ]
    ]);

eq("ZIP", Bs_List.zip(/* [] */0, /* :: */[
          1,
          /* [] */0
        ]), /* [] */0);

eq("ZIP", Bs_List.zip(/* [] */0, /* [] */0), /* [] */0);

eq("ZIP", Bs_List.zip(/* :: */[
          1,
          /* :: */[
            2,
            /* :: */[
              3,
              /* [] */0
            ]
          ]
        ], /* [] */0), /* [] */0);

eq("ZIP", Bs_List.zip(/* :: */[
          1,
          /* :: */[
            2,
            /* :: */[
              3,
              /* [] */0
            ]
          ]
        ], /* :: */[
          2,
          /* :: */[
            3,
            /* :: */[
              4,
              /* [] */0
            ]
          ]
        ]), /* :: */[
      /* tuple */[
        1,
        2
      ],
      /* :: */[
        /* tuple */[
          2,
          3
        ],
        /* :: */[
          /* tuple */[
            3,
            4
          ],
          /* [] */0
        ]
      ]
    ]);

function mod2(x) {
  return +(x % 2 === 0);
}

eq("PARTITION", Bs_List.partition(mod2, /* :: */[
          1,
          /* :: */[
            2,
            /* :: */[
              3,
              /* :: */[
                2,
                /* :: */[
                  3,
                  /* :: */[
                    4,
                    /* [] */0
                  ]
                ]
              ]
            ]
          ]
        ]), /* tuple */[
      /* :: */[
        2,
        /* :: */[
          2,
          /* :: */[
            4,
            /* [] */0
          ]
        ]
      ],
      /* :: */[
        1,
        /* :: */[
          3,
          /* :: */[
            3,
            /* [] */0
          ]
        ]
      ]
    ]);

eq("PARTITION", Bs_List.partition(mod2, /* :: */[
          2,
          /* :: */[
            2,
            /* :: */[
              2,
              /* :: */[
                4,
                /* [] */0
              ]
            ]
          ]
        ]), /* tuple */[
      /* :: */[
        2,
        /* :: */[
          2,
          /* :: */[
            2,
            /* :: */[
              4,
              /* [] */0
            ]
          ]
        ]
      ],
      /* [] */0
    ]);

eq("PARTITION", Bs_List.partition((function (x) {
            return 1 - mod2(x);
          }), /* :: */[
          2,
          /* :: */[
            2,
            /* :: */[
              2,
              /* :: */[
                4,
                /* [] */0
              ]
            ]
          ]
        ]), /* tuple */[
      /* [] */0,
      /* :: */[
        2,
        /* :: */[
          2,
          /* :: */[
            2,
            /* :: */[
              4,
              /* [] */0
            ]
          ]
        ]
      ]
    ]);

eq("SPLIT", Bs_List.split(/* [] */0), /* tuple */[
      /* [] */0,
      /* [] */0
    ]);

eq("SPLIT", Bs_List.split(/* :: */[
          /* tuple */[
            1,
            2
          ],
          /* [] */0
        ]), /* tuple */[
      /* :: */[
        1,
        /* [] */0
      ],
      /* :: */[
        2,
        /* [] */0
      ]
    ]);

eq("SPLIT", Bs_List.split(/* :: */[
          /* tuple */[
            1,
            2
          ],
          /* :: */[
            /* tuple */[
              3,
              4
            ],
            /* [] */0
          ]
        ]), /* tuple */[
      /* :: */[
        1,
        /* :: */[
          3,
          /* [] */0
        ]
      ],
      /* :: */[
        2,
        /* :: */[
          4,
          /* [] */0
        ]
      ]
    ]);

eq("FILTER", Bs_List.filter(mod2, /* :: */[
          1,
          /* :: */[
            2,
            /* :: */[
              3,
              /* :: */[
                4,
                /* [] */0
              ]
            ]
          ]
        ]), /* :: */[
      2,
      /* :: */[
        4,
        /* [] */0
      ]
    ]);

eq("FILTER", Bs_List.filter(mod2, /* :: */[
          1,
          /* :: */[
            3,
            /* :: */[
              41,
              /* [] */0
            ]
          ]
        ]), /* [] */0);

eq("FILTER", Bs_List.filter(mod2, /* [] */0), /* [] */0);

eq("FILTER", Bs_List.filter(mod2, /* :: */[
          2,
          /* :: */[
            2,
            /* :: */[
              2,
              /* :: */[
                4,
                /* :: */[
                  6,
                  /* [] */0
                ]
              ]
            ]
          ]
        ]), /* :: */[
      2,
      /* :: */[
        2,
        /* :: */[
          2,
          /* :: */[
            4,
            /* :: */[
              6,
              /* [] */0
            ]
          ]
        ]
      ]
    ]);

function id(x) {
  return x;
}

eq("MAP", Bs_List.map(Bs_List.init(5, id), (function (x) {
            return (x << 1);
          })), /* :: */[
      0,
      /* :: */[
        2,
        /* :: */[
          4,
          /* :: */[
            6,
            /* :: */[
              8,
              /* [] */0
            ]
          ]
        ]
      ]
    ]);

eq("MAP", Bs_List.map(/* [] */0, id), /* [] */0);

eq("MAP", Bs_List.map(/* :: */[
          1,
          /* [] */0
        ], (function (x) {
            return -x | 0;
          })), /* :: */[
      -1,
      /* [] */0
    ]);

function add(a, b) {
  return a + b | 0;
}

var a = Bs_List.init(10, id);

var b$1 = Bs_List.init(10, id);

var c = Bs_List.init(8, id);

var d = Bs_List.init(10, (function (x) {
        return (x << 1);
      }));

eq("MAP2", Bs_List.map2(add, a, b$1), d);

eq("MAP2", Bs_List.map2(add, /* [] */0, /* :: */[
          1,
          /* [] */0
        ]), /* [] */0);

eq("MAP2", Bs_List.map2(add, /* :: */[
          1,
          /* [] */0
        ], /* [] */0), /* [] */0);

eq("MAP2", Bs_List.map2(add, /* [] */0, /* [] */0), /* [] */0);

eq("MAP2", Bs_List.map2(add, a, b$1), Bs_List.append(Bs_List.map(c, (function (x) {
                return (x << 1);
              })), /* :: */[
          16,
          /* :: */[
            18,
            /* [] */0
          ]
        ]));

Mt.from_pair_suites("bs_link_list_test.ml", suites[0]);

var N = 0;

var A = 0;

exports.suites  = suites;
exports.test_id = test_id;
exports.eq      = eq;
exports.b       = b;
exports.N       = N;
exports.A       = A;
exports.mod2    = mod2;
exports.id      = id;
exports.add     = add;
/* u Not a pure module */

File: jscomp/test/bs_link_list_test.ml
let suites :  Mt.pair_suites ref  = ref []
let test_id = ref 0
let eq loc x y = 
  incr test_id ; 
  suites := 
    (loc ^" id " ^ (string_of_int !test_id), (fun _ -> Mt.Eq(x,y))) :: !suites

let b loc x  = 
  incr test_id ; 
  suites := 
    (loc ^" id " ^ (string_of_int !test_id), 
     (fun _ -> Mt.Ok x)) :: !suites

(* module N = Bs.LinkList     *)
module N = Bs.List
module A = Bs.Array 

let () = 
  let u = (N.init 5 (fun[@bs] i -> i * i )) in 

  (* N.checkInvariant u ; *)
  let f i = 
    eq __LOC__ (N.nthAssert u i) (i * i) in 
  for i = 0 to 4 do 
    f i 
  done ;
  eq __LOC__  (N.map u (fun [@bs] i -> i + 1)) [1;2;5;10;17]

let () =
  let (=~) = eq "FLATTEN" in 

  
    N.(flatten
        [[1]; [2]; [3];[]; init 4 (fun [@bs] i -> i )]
      ) =~
    [1;2;3; 0;1;2;3];
  N.flatten [] =~ [];  
  N.flatten [[];[]; [2]; [1];[2];[]] =~ [2;1;2]
    
let () = 
  eq __LOC__
    (N.
       (append 
          (init 100 (fun [@bs] i -> i) )
          (init 100 (fun [@bs] i -> i)))
     |> N.toArray
    )

    (A.
       (append 
          (init 100 (fun [@bs] i -> i) )
          (init 100 (fun [@bs] i -> i)))
    )
let () = 
  let (=~) = eq "APPEND" in 
  N.append [1] [] =~ [1];
  N.append [] [1] =~ [1]

let () =     
  let (=~) = eq "ZIP"   in 
  
  (N.zip [1;2;3] [3;4]) =~ [1,3; 2,4];
  N.zip [] [1] =~ [];
  N.zip [] [] =~ [];
  N.zip [1;2;3] [] =~ [] ;
  N.zip [1;2;3] [2;3;4] =~ [1,2;2,3;3,4]
let mod2 = (fun[@bs] x -> x mod 2 = 0) 
let () =   
  let (=~) = eq "PARTITION" in 
  
  (N.partition [1;2;3;2;3;4] mod2 )
  =~ ([2;2;4], [1;3;3]);
  (N.partition [2;2;2;4] mod2)
  =~ ([2;2;2;4], []);
  (N.partition  [2;2;2;4] (fun[@bs] x -> not (mod2 x [@bs] )))
  =~ ([], [2;2;2;4])

let () =   
  let (=~) = eq "UNZIP" in 
  N.unzip [] =~ ([],[]) ; 
  N.unzip [1,2] =~ ([1] ,[2]);
  N.unzip [1,2;3,4] =~ ([1;3], [2;4])

let () = 
  let (=~) = eq "FILTER" in 
  N.filter [1;2;3;4] mod2 =~ [2;4];
  N.filter [1;3;41] mod2 =~ [];
  N.filter [] mod2 =~ [];
  N.filter  [2;2;2;4;6] mod2 =~ [2;2;2;4;6]
let id : int -> int [@bs] = fun [@bs] x -> x 

let () =   
  let (=~) = eq "MAP" in 
  N.map (N.init 5 id )(fun [@bs] x -> x * 2 ) 
  =~ [0;2;4;6;8];
  N.map [] id  =~ [];
  N.map [1] (fun [@bs] x-> -x)  =~ [-1]
let add = (fun [@bs] a b -> a + b)
let length_10_id = N.init 10 id  
let length_8_id = N.init 8 id 
let () = 
  let (=~) = eq "MAP2" in   
  let b = length_10_id in
  let c = length_8_id in 
  let d = N.init 10 (fun [@bs] x -> 2 * x ) in     
  let map2_add x y = N.map2  x y add in 
  map2_add length_10_id b =~ d ;
  map2_add [] [1] =~ [];
  map2_add [1] [] =~ [];
  map2_add [] [] =~ [];
  map2_add length_10_id b =~  N.(append (map c (fun[@bs] x -> x * 2)) [16;18]);
  map2_add length_10_id length_8_id =~
  N.(mapi length_8_id (fun [@bs] i x -> i + x ) )
let () =   
  let (=~) = eq "TAKE" in 
  N.takeOpt [1;2;3] 2 =~ Some [1;2];
  N.takeOpt [] 1 =~ None;
  N.takeOpt [1;2] 3 =~ None ; 
  N.takeOpt [1;2] 2 =~ Some [1;2];
  N.takeOpt length_10_id 8 =~ Some length_8_id ;
  N.takeOpt length_10_id 0 =~ Some [];
  N.takeOpt length_8_id (-2) =~ None 

let () =   
  let (=~) = eq "DROP" in 
  N.dropOpt length_10_id 10 =~ Some [];
  N.dropOpt length_10_id 8 =~ Some [8;9];
  N.dropOpt length_10_id 0 =~ Some length_10_id ;
  N.dropOpt length_8_id (-1) =~ None

let () = 
  let (=~) = eq "SPLIT" in 
  let a = N.init 5 id in 
  N.splitAtOpt [] 1 =~ None;
  N.splitAtOpt a 6 =~ None;
  N.splitAtOpt a 5 =~ Some (a,[]);
  N.splitAtOpt a 4 =~ Some ([0;1;2;3],[4]);
  N.splitAtOpt a 3 =~ Some ([0;1;2],[3;4]);
  N.splitAtOpt a 2 =~ Some ([0;1],[2;3;4]);
  N.splitAtOpt a 1 =~ Some ([0],[1;2;3;4]);
  N.splitAtOpt a 0 =~ Some ([],a);
  N.splitAtOpt a (-1) =~ None
let succx =   (fun[@bs] x -> x + 1)
let ()   = 
  
  eq __LOC__ N.(headOpt length_10_id, tailOpt length_10_id)  (Some 0, N.dropOpt length_10_id 1);
  eq __LOC__ (N.headOpt [])  None ;
  N.iteri length_10_id (fun[@bs] i x ->
     eq __LOC__ (N.nthOpt length_10_id i) (Some x));
  eq __LOC__ (N.nthOpt  length_10_id (-1) ) None; 
  eq __LOC__ (N.nthOpt  length_10_id 12 ) None;
  eq __LOC__ (N.init 0 id) [];
  eq __LOC__ (N.(rev (rev length_10_id)))  length_10_id ; 
  eq __LOC__ (N.(rev (rev length_8_id)))  length_8_id ;
  eq __LOC__ (N.rev []) [];
  eq __LOC__ 
  (N.rev (N.mapRev length_10_id succx))
  (N.map length_10_id succx  );
  eq __LOC__
  (N.foldLeft length_10_id 0 add) 45;
  eq __LOC__
  (N.foldRight length_10_id 0 add) 45;
  (* eq __LOC__ 
  (N.mapRev2 length_10_id length_8_id add ) *)

  
;; Mt.from_pair_suites __FILE__ !suites
